### ISSUE TITLE
Throttler overview

### ISSUE THREAD

**Issue by hahahannes** [rucio/rucio#2542]
Motivation
----------
There are multiple throttler tickets for different use cases:
FIFO:
- [x] throttle per activity -  source RSE #2611 
- [x] throttle per activity -  destination RSE
- [ ] throttle per activity -  destination and source RSE #53 
- [x] throttle per all activities -  destination RSE #1836 
- [x] throttle per all activities -  source RSE #2611 
- [ ] throttle per all activities -  source and destination RSE #53 

Grouped FIFO:
- [x] throttle per all activities -  destination RSE #2220 
- [x] throttle per all activities -  source RSE #2611
- [ ] throttle per all activities -  destination and source RSE #53 

Modification
------------

**Comment by hahahannes** [comment]
There should be also an easy way to configure the throttler in the config table .
For example there could be a throttler mode in the config like `throttler_mode`:
- throttler_mode, SRC_PER_ACT (to throttle per source RSE and per activity)
- throttler_mode, DEST_PER_ACT
- throttler_mode, SRC_DEST_PER_ACT
- throttler_mode, SRC_ALL_ACT
- throttler_mode, DEST_ALL_ACT
- throttler_mode, SRC_DEST_ALL_ACT (to throttle per source and destination RSE and over all activites)

Then each RSE would need a threshold value in combination with the mode to allow switching between modes like:
- 'All Activities, DEST_ALL_ACT, MOCK', 10000
- 'User Subscription, SRC_DEST_ALL_ACT, MOCK', 20000

This would add the modes to values in the rse_transfer_limits table.

Also each RSE should be configured to use FIFO or grouped FIFO depending on the mode:
- 'DEST_ALL_ACT, MOCK', FIFO

**Comment by bari12** [comment]
Yes, we would need some kind of mode specifier in the table. I wonder if this should be an additional column though.

**Comment by hahahannes** [comment]
Also there are some special cases to be considered
- if the throttler uses destination and source, it could be that the destination RSE is configured to use FIFO and the source RSE grouped FIFO, then its not clear what to use
- if the throttler uses destination and source, it could be that a RSE is used as source and destination, then the configuration would also need a field to specify that like `'All Activities, DEST_ALL_ACT, MOCK, DEST', 10000` if MOCK is used as destination RSE

**Comment by hahahannes** [comment]
> Yes, we would need some kind of mode specifier in the table. I wonder if this should be an additional column though.

@bari12 You mean in the rse_transfer_limits table? I thought that the throttler can only have one mode, so I would set the value only once

**Comment by bari12** [comment]
Yes, in `rse_transfer_limits`. It's just not super nice that all the "configuration" is part of a long string. For some things it's not possible otherwise, but we should try to put as many things as possible into separate columns.

**Comment by hahahannes** [comment]
Ah yes makes sense. We should be maybe also check why the throttler thresholds are stored in the config table. I think the ones in the `rse_transfer_limits` table are unused and it would be better to use them and then to use seperate colums, like you said

**Comment by hahahannes** [comment]
I think this schema for `rse_transfer_limits` would fit.

| RSE ID        | Activity           | Max transfers  | Transfers | Waiting | Direction | Mode | Volumn | Deadline
| ------------- |:-------------:| -----:|-----:|-----:| -----:|  -----:| -----:| -----:|
| MOCK      | ALL | 10000 | 10 | 10 | DEST | FIFO | 10000 | -
| MOCK      | User Subscription      |   500 | 10 | 10 | SRC | FIFO | - | -
| MOCK      | User Subscription      |   500 | 10 | 10 | DEST | GFIFO | 500 | -
| MOCK      | User Subscription      |   500 | 10 | 10 | SRC | GFIFO | - | 3


Then depending on the throttler mode the thresholds and the release mode can be read out, e.g the mode `DEST_PER_ACT` would read the third row, `DEST_ALL_ACT` would read the first row and `SRC_PER_ACT ` would read the second row

**Comment by bari12** [comment]
Yes, I think that would work. (Although Volum**e** instead of Volum**n** 😄 

Right now there is also a column `rse_expression` in the table. I am not quite sure why, maybe this was just put there in preparation when the throttler was created.

**Comment by bari12** [comment]
In the new implementation the throttler lost the ability to set a limit for `all_rses`, only limits for individual rses are possible now. This needs to be added again.

**Comment by bari12** [comment]
I am closing all the throttler stuff; We will have a new ticket describing the changes needed.
